### PR TITLE
Performance improvement for io::inputbuffer

### PR DESCRIPTION
**Pull request by snnn**
Do not add chars to string one by one.
In most cases, buffer size >= max line length. 
Let's do it in one function call.

### CODE REVIEW

**Comment by snnn** [comment]
For input like:
"aaa\rbbb\r\n"

The old code will output: "aaabbb"
This new code will output: "aaa\rbbb"

I think this behavior change is acceptable. Do you?

**Comment by mrry** [comment]
Have you measured the performance improvement?

It certainly *looks* more efficient to me, but I'd like to weigh this against the cost of making a (probably irrelevant, but you never know...) backwards incompatible change.

**Comment by snnn** [comment]
Working. I'll post the result later.

**Comment by snnn** [comment]
input data: A text file in libsvm format, 10906667862 bytes, 164540416 lines, with CRLF line terminators

test code 1:
```cpp
#include <mutex>
#include <Windows.h>
#include <stdio.h>

#include "tensorflow/core/platform/env.h"
#include "tensorflow/core/lib/core/errors.h"
#include "tensorflow/core/lib/io/inputbuffer.h"
using namespace tensorflow;

int test1(const char* filename) {
	LARGE_INTEGER li;
	QueryPerformanceFrequency(&li);
	Env* env = Env::Default();
	std::unique_ptr<RandomAccessFile> file_;
	const ::tensorflow::Status status = env->NewRandomAccessFile(filename, &file_);
	if (!status.ok()) {
		return -1;
	}
	LARGE_INTEGER startTime, endTime;
	int kBufferSize = 2 * 1024 * 1024;
	io::InputBuffer* input_buffer = new io::InputBuffer(file_.get(), kBufferSize);
	QueryPerformanceCounter(&startTime);
	string line_contents;
	while (true) {
		Status status = input_buffer->ReadLine(&line_contents);
		if (errors::IsOutOfRange(status)) {
			break;
		}
		if (!status.ok()) {
			return -1;
		}
	}
	QueryPerformanceCounter(&endTime);
	const double seconds = ((endTime.QuadPart - startTime.QuadPart) / (double)li.QuadPart);
	printf("dur = %g\n", seconds);
	return 0;
}
```
test code 2:
Almost the same as test code 1, with only one line change:
```
while (true) {
    string line_contents;  // this line is moved from the line before "while(true)"
    Status status = input_buffer->ReadLine(&line_contents);
}
```
Result:

<table>
<thead>
<tr>
<td>&nbsp;</td>
<td>Before</td>
<td>After</td>
<td>Changes</td>
</tr>
</thead>
<tbody>
<tr>
<td>test1</td>
<td>67 seconds</td>
<td>26 seconds</td>
<td> - 61%</td>
</tr>
<tr>
<td>test2(cache line_contents)</td>
<td>34 seconds</td>
<td>19 seconds</td>
<td> - 44%</td>
</tr>
</tbody>
</table>

**Comment by mrry** [comment]
Great, thanks for confirming that. I'm satisfied with the change, so let's run the tests and merge if they pass.

@tensorflow-jenkins test this please.

**Review by mrry — APPROVED** [review]

File: tensorflow/core/lib/io/inputbuffer.cc [hunk 1]
```diff
@@ -43,25 +43,26 @@ Status InputBuffer::FillBuffer() {
...
       return Status::OK();
     }
-    // We don't append '\r' to *result
-    if (c != '\r') {
-      *result += c;
-    }
+    if (buf_remain > 0) result->append(pos_, buf_remain);
+    // Get more data into buffer
+    s = FillBuffer();
+    assert(pos_ == buf_);
```
Comment (mrry) [inline tensorflow/core/lib/io/inputbuffer.cc]: One minor change, sorry: can you replace this line with `DCHECK_EQ(pos_, buf_);` to match the style of the rest of the codebase? Thanks!

**Review by mrry — CHANGES_REQUESTED** [review]
Just noticed one minor change... should be good to merge after changing.

File: tensorflow/core/lib/io/inputbuffer.cc (hunk 1 above)
Comment (snnn) [inline tensorflow/core/lib/io/inputbuffer.cc]: Sure.

Comment (snnn) [inline tensorflow/core/lib/io/inputbuffer.cc]: Done.

**Review by mrry — APPROVED** [review]
Great, thanks! Looks good to me.

**Comment by mrry** [comment]
@tensorflow-jenkins test this please.

**Pull request merged**